session.go

```diff
@@ -0,0 +1,309 @@
+package wsep
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// Session represents a `screen` session.
+type Session struct {
+	// close receives nil when the session should be closed.
+	close chan struct{}
+	// closing receives nil when the session has begun closing.  The underlying
+	// process may still be exiting.  closing implies ready.
+	closing chan struct{}
+	// command is the original command used to spawn the session.
+	command *Command
+	// configFile is the location of the screen configuration file.
+	configFile string
+	// done receives nil when the session has completely shut down and the process
+	// has exited.  done implies closing and ready.
+	done chan struct{}
+	// error hold any error that occurred while starting the session.
+	error error
+	// execer is used to spawn the session and ready commands.
+	execer Execer
+	// id holds the id of the session for both creating and attaching.
+	id string
+	// options holds options for configuring the session.
+	options *Options
+	// ready receives nil when the session is ready to be attached.  error must be
+	// checked after receiving on this channel.
+	ready chan struct{}
+	// socketsDir is the location of the directory where screen should put its
+	// sockets.
+	socketsDir string
+	// timer will close the session when it expires.
+	timer *time.Timer
+}
+
+// NewSession creates and immediately starts a new session.  Any errors with
+// starting are returned on Attach().  The session will close itself if nothing
+// is attached for the duration of the session timeout.
+func NewSession(id string, command *Command, execer Execer, options *Options) *Session {
+	tempdir := filepath.Join(os.TempDir(), "coder-screen")
+	s := &Session{
+		close:      make(chan struct{}),
+		closing:    make(chan struct{}),
+		command:    command,
+		configFile: filepath.Join(tempdir, "config"),
+		done:       make(chan struct{}),
+		execer:     execer,
+		options:    options,
+		ready:      make(chan struct{}),
+		socketsDir: filepath.Join(tempdir, "sockets"),
+	}
+	go s.lifecycle(id)
+	return s
+}
+
+// lifecycle manages the lifecycle of the session.
+func (s *Session) lifecycle(id string) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Close the session after a timeout.  Timeouts created with AfterFunc can be
+	// reset; we will reset it as long as there are active connections.
+	s.timer = time.AfterFunc(s.options.SessionTimeout, s.Close)
+	defer s.timer.Stop()
+
+	// Close the session down immediately if there was an error and store that
+	// error to emit on Attach() calls.
+	process, err := s.start(ctx, id)
+	if err != nil {
+		s.error = err
+		close(s.ready)
+		close(s.closing)
+		close(s.done)
+		return
+	}
+
+	// Mark the session as fully done when the process exits.
+	go func() {
+		err = process.Wait()
+		if err != nil {
+			s.error = err
```

Comment:
I don't think it's safe to write to s.error from two different goroutines as you have no mutex or other sync primitive protecting it.